jgrapht-core/src/main/java/org/jgrapht/alg/color/ColorRefinementAlgorithm.java

```diff
@@ -68,279 +78,181 @@ public ColorRefinementAlgorithm(Graph<V, E> graph, Coloring<V> alpha) {
-            if(C.get(c).size() != A.get(c).size()) {
-                minColorDegree[c] = 0;
+        // update minColorDegree
+        for (Integer c : adjacentColors) {
+            // if there is a vertex with colorDegree(v) = 0 < 1, set minimum color degree to
+            // 0
+            if (rep.colorClasses.get(c).size() != rep.positiveDegreeColorClasses.get(c).size()) {
+                rep.minColorDegree[c] = 0;
             } else {
-                minColorDegree[c] = maxColorDegree[c];
-                for(V v : A.get(c)) {
-                    if (colorDegree.get(v) < minColorDegree[c]) {
-                        minColorDegree[c] = colorDegree.get(v);
+                rep.minColorDegree[c] = rep.maxColorDegree[c];
+                for (V v : rep.positiveDegreeColorClasses.get(c)) {
+                    if (rep.colorDegree.get(v) < rep.minColorDegree[c]) {
+                        rep.minColorDegree[c] = rep.colorDegree.get(v);
                     }
                 }
             }
         }
+
+        return adjacentColors;
     }
 
     /**
-     * Helper method for getColoring().
-     * Partition the colors that do not have the same minimum and maximum color degree. That is, these colors are not completely refined.
+     * Helper method that cleanups the internal representation of color degrees for a new iteration.
      *
-     * @param color the color mapping
-     * @param C the mapping from all colors to their classes
-     * @param A the mapping from all colors to their classes with colorDegree(v) >= 1
-     * @param refineStack the stack containing all colors that have to be refined
-     * @param adjacentColors the list of all colors that have at least one vertex with colorDegree >= 1
-     * @param maxColorDegree the mapping from color to its maximum color degree
-     * @param minColorDegree the mapping from color to its minimum color degree
-     * @param colorDegree the mapping from vertex to the color degree (number of neighbors with different colors) of the vertex
+     * @param adjacentColors the list of all colors that have at least one vertex with colorDegree
+     *        >= 1
+     * @param rep the coloring representation
      */
-    private void calculateColorPartition(Map<V, Integer> color, HashMap<Integer, List<V>> C, HashMap<Integer, List<V>> A, Deque<Integer> refineStack, ArrayList<Integer> adjacentColors, int[] maxColorDegree, int[] minColorDegree, Map<V, Integer> colorDegree) {
-        // subset of adjacentColors that will be split up into different color classes
-        ArrayList<Integer> Colors_split = new ArrayList<>();
-
-        for(Integer c : adjacentColors) {
-            if(minColorDegree[c] < maxColorDegree[c]) {
-                Colors_split.add(c);
+    private void cleanupColorDegrees(Set<Integer> adjacentColors, ColoringRepresentation rep)
+    {
+        adjacentColors.stream().forEach(c -> {
+            for (V v : rep.positiveDegreeColorClasses.get(c)) {
```

Comment:
Consider normalising this to use a `.forEach()`, or to make the outer `.stream().forEach()` just be a for-loop as well.